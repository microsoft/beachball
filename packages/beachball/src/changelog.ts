import { PackageInfo } from './bump';
import path from 'path';
import fs from 'fs';
import { generateTag } from './tag';
import { ChangeInfo, ChangeSet } from './ChangeInfo';

interface ChangelogEntry {
  comment: string;
  author: string;
  commit: string;
}

interface PackageChangelog {
  name: string;
  date: Date;
  version: string;
  comments: {
    prerelease?: ChangelogEntry[];
    patch?: ChangelogEntry[];
    minor?: ChangelogEntry[];
    major?: ChangelogEntry[];
    none?: ChangelogEntry[];
  };
}

interface ChangelogJsonEntry {
  date: string;
  version: string;
  tag: string;
  comments: {
    prerelease?: ChangelogEntry[];
    patch?: ChangelogEntry[];
    minor?: ChangelogEntry[];
    major?: ChangelogEntry[];
    none?: ChangelogEntry[];
  };
}

interface ChangelogJson {
  name: string;
  entries: ChangelogJsonEntry[];
}

export function getPackageChangelogs(changeSet: ChangeSet, packageInfos: { [pkg: string]: PackageInfo }) {
  const changelogs: { [pkgName: string]: PackageChangelog } = {};

  for (let [_, change] of changeSet) {
    const { packageName } = change;

    changelogs[packageName] = changelogs[packageName] || {
      name: packageName,
      version: packageInfos[packageName].version,
      date: new Date(),
    };

    changelogs[packageName].comments = changelogs[packageName].comments || {};
    changelogs[packageName].comments[change.type] = changelogs[packageName].comments[change.type] || [];
    changelogs[packageName].comments[change.type]!.push({
      comment: change.comment,
      author: change.email,
      commit: change.commit,
    });
  }

  return changelogs;
}

export function writeChangelog(changeSet: ChangeSet, packageInfos: { [pkg: string]: PackageInfo }) {
  const changelogs = getPackageChangelogs(changeSet, packageInfos);

  Object.keys(changelogs).forEach(pkg => {
    const packagePath = path.dirname(packageInfos[pkg].packageJsonPath);

    if (
      changelogs[pkg].comments.major ||
      changelogs[pkg].comments.minor ||
      changelogs[pkg].comments.patch ||
      changelogs[pkg].comments.prerelease
    ) {
      const changelogFile = path.join(packagePath, 'CHANGELOG.md');
      const previousContent = fs.existsSync(changelogFile) ? fs.readFileSync(changelogFile).toString() : '';
      const nextContent = renderChangelog(previousContent, changelogs[pkg]);
      fs.writeFileSync(changelogFile, nextContent);
    }

    try {
      const changelogJsonFile = path.join(packagePath, 'CHANGELOG.json');
      const previousJson = fs.existsSync(changelogJsonFile)
        ? JSON.parse(fs.readFileSync(changelogJsonFile).toString())
        : { entries: [] };
      const nextJson = renderJsonChangelog(previousJson, changelogs[pkg]);
      fs.writeFileSync(changelogJsonFile, JSON.stringify(nextJson, null, 2));
    } catch (e) {
      console.warn('The CHANGELOG.json file is invalid, skipping writing to it', e);
    }
  });
}

function renderJsonChangelog(previous: ChangelogJson, changelog: PackageChangelog) {
  const result: ChangelogJson = {
    name: changelog.name,
    entries: [...previous.entries] || [],
  };

  const newEntry: ChangelogJsonEntry = {
    date: changelog.date.toUTCString(),
    tag: generateTag(changelog.name, changelog.version),
    version: changelog.version,
    comments: changelog.comments,
  };

  result.entries.unshift(newEntry);

  return result;
}

function renderChangelog(previous: string, changelog: PackageChangelog) {
  const previousLogEntries = previous ? '\n' + previous.substring(previous.indexOf('##')) : '';

  return (
    `# Change Log - ${changelog.name}\n\n` +
    `This log was last generated on ${changelog.date.toUTCString()} and should not be manually modified.\n` +
    renderPackageChangelog(changelog) +
    previousLogEntries
  );
}

function renderPackageChangelog(changelog: PackageChangelog) {
  return (
    `\n## ${changelog.version}\n` +
    `${changelog.date.toUTCString()}\n` +
    (changelog.comments.major
      ? '\n### Major\n\n' + changelog.comments.major.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '') +
    (changelog.comments.minor
      ? '\n### Minor changes\n\n' +
        changelog.comments.minor.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '') +
    (changelog.comments.patch
      ? '\n### Patches\n\n' +
        changelog.comments.patch.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '') +
    (changelog.comments.prerelease
      ? '\n### Changes\n\n' +
        changelog.comments.prerelease.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '')
  );
}
