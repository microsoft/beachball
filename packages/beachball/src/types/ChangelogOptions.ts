import { ChangelogJson, PackageChangelog, ChangelogEntry } from './ChangeLog';
import { ChangeType } from './ChangeInfo';

/**
 * Options for changelog-related configuration.
 *
 * If you would like to entirely customize rendering, use `renderPackageChangelog`.
 * Otherwise, you can provide as many or as few of the other custom renderer options as you'd like
 * to achieve the desired level of customization.
 */
export interface ChangelogOptions {
  /** Options for grouping packages together in a single changelog. */
  groups?: ChangelogGroupOptions[];

  /**
   * Use this for full custom rendering of the entire changelog markdown for a particular package version.
   * Default renderers (and `customRenderers` if provided) will be available in `renderInfo.renderers`
   * but will not be called automatically.
   */
  renderPackageChangelog?: (renderInfo: PackageChangelogRenderInfo) => Promise<string>;

  /**
   * Fine-grained custom renderers for individual parts of the changelog.
   * If using a custom `renderPackageChangelog`, these will not be called automatically.
   */
  customRenderers?: ChangelogRenderers;
}

/**
 * Options for generating a changelog for a group of packages.
 */
export interface ChangelogGroupOptions {
  /**
   * The main package which a group of changes bubbles up to.
   * All changes within the group are used to describe changes for the master package.
   */
  masterPackageName: string;

  /** minimatch pattern (or array of minimatch) to detect which packages should be included in this group */
  include: string | string[];

  /** minimatch pattern (or array of minimatch) to detect which packages should be excluded in this group */
  exclude?: string | string[];

  changelogPath: string;
}

/**
 * Info used for rendering the changelog markdown for a particular package version.
 */
export interface PackageChangelogRenderInfo {
  /** Existing json in CHANGELOG.json.  */
  previousJson: ChangelogJson | undefined;

  /** Changelog for a package version that is going to be added to full changelog. */
  newVersionChangelog: PackageChangelog;

  /** True if this is a grouped changelog. */
  isGrouped: boolean;

  /**
   * Renderers for individual elements of the changelog.
   * If any custom renderers were provided in `ChangelogOptions.customRenderers`, they will be included here.
   * Default renderers will be included in cases where a custom option wasn't provided.
   */
  renderers: Required<ChangelogRenderers>;
}

export interface ChangelogRenderers {
  /**
   * Custom renderer for the header for a particular package version.
   *
   * Default is like this (no leading or trailing newlines):
   * ```txt
   * ## 1.23.1
   * Wed, 25 Mar 2020 20:20:02 GMT
   * ```
   */
  renderHeader?: (renderInfo: PackageChangelogRenderInfo) => Promise<string>;

  /**
   * Custom renderer for the section about `changeType` changes for a particular package version.
   *
   * Default is like this (no leading or trailing newlines):
   * ```txt
   * ### Minor changes
   *
   * - Really interesting change (user1@microsoft.com)
   * ```
   */
  renderChangeTypeSection?: (changeType: ChangeType, renderInfo: PackageChangelogRenderInfo) => Promise<string>;

  /**
   * Custom renderer for the section header about `changeType` changes for a particular package version.
   *
   * Default is like this (no leading or trailing newlines):
   * ```txt
   * ### Minor changes
   * ```
   */
  renderChangeTypeHeader?: (changeType: ChangeType, renderInfo: PackageChangelogRenderInfo) => Promise<string>;

  /**
   * Custom renderer for the list of `changeType` changes (not including the change type header)
   * for a particular package version.
   *
   * Default is like this for non-grouped changelogs (no leading or trailing newlines):
   * ```txt
   * - Really interesting change (user1@microsoft.com)
   * - Boring change (user2@microsoft.com)
   * ```
   *
   * Or like this for grouped changelogs:
   * ```txt
   * - `foo`
   *   - Really interesting change (user1@microsoft.com)
   * - `bar`
   *   - Boring change (user2@microsoft.com)
   * ```
   */
  renderEntries?: (changeType: ChangeType, renderInfo: PackageChangelogRenderInfo) => Promise<string>;

  /**
   * Custom renderer for an individual change entry.
   *
   * Default is like this (no leading or trailing newlines):
   * ```txt
   * - Really interesting change (user1@microsoft.com)
   * ```
   */
  renderEntry?: (entry: ChangelogEntry, renderInfo: PackageChangelogRenderInfo) => Promise<string>;
}
