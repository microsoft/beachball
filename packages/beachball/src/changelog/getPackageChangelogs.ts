import { ChangeSet } from '../types/ChangeInfo';
import { PackageInfo } from '../types/PackageInfo';
import { PackageChangelog } from '../types/ChangeLog';
export function getPackageChangelogs(changeSet: ChangeSet, packageInfos: {
  [pkg: string]: PackageInfo;
}) {
  const changelogs: {
    [pkgName: string]: PackageChangelog;
  } = {};
  for (let [_, change] of changeSet) {
    const { packageName } = change;
    changelogs[packageName] = changelogs[packageName] || {
      name: packageName,
      version: packageInfos[packageName].version,
      date: new Date(),
    };
    changelogs[packageName].comments = changelogs[packageName].comments || {};
    changelogs[packageName].comments[change.type] = changelogs[packageName].comments[change.type] || [];
    changelogs[packageName].comments[change.type]!.push({
      comment: change.comment,
      author: change.email,
      commit: change.commit,
    });
  }
  return changelogs;
}
