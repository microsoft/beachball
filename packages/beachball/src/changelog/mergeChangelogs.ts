import _ from 'lodash';
import { PackageChangelog } from '../types/ChangeLog';
import { PackageInfo } from '../types/PackageInfo';

/**
 * Merge multiple PackageChangelog into one.
 * `name`, `date` and `version` will be using the values from master changelog. `comments` are merged.
 */
export function mergeChangelogs(
  changelogs: PackageChangelog[],
  masterPackage: PackageInfo
): PackageChangelog | undefined {
  if (changelogs.length < 1 || !masterPackage) {
    return undefined;
  }

  const result: PackageChangelog = {
    name: masterPackage.name,
    version: masterPackage.version,
    date: new Date(),
    comments: {},
  };

  changelogs.forEach(changelog => {
    Object.keys(changelog.comments).forEach(changeType => {
      if (changelog.comments[changeType]) {
        result.comments[changeType] = (result.comments[changeType] || []).concat(changelog.comments[changeType]);
      }
    });
  });

  return result;
}
