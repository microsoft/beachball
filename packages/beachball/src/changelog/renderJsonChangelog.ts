import { generateTag } from '../tag';
import { PackageChangelog, ChangelogJson, ChangelogJsonEntry } from '../types/ChangeLog';

export function renderJsonChangelog(changelog: PackageChangelog, previousChangelog: ChangelogJson | undefined) {
  const result: ChangelogJson = {
    name: changelog.name,
    entries: previousChangelog?.entries ? [...previousChangelog.entries] : [],
  };
  const newEntry: ChangelogJsonEntry = {
    date: changelog.date.toUTCString(),
    tag: generateTag(changelog.name, changelog.version),
    version: changelog.version,
    comments: changelog.comments,
  };
  result.entries.unshift(newEntry);
  return result;
}
