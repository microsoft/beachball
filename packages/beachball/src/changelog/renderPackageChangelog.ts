import { ChangelogEntry } from '../types/Changelog';
import _ from 'lodash';
import { PackageChangelogRenderInfo, ChangelogRenderers } from '../types/ChangelogOptions';
import { ChangeType } from '../types/ChangeInfo';

const groupNames: { [k in ChangeType]: string } = {
  major: 'Major changes',
  minor: 'Minor changes',
  patch: 'Patches',
  prerelease: 'Changes',
  none: '', // not used
};

export const defaultRenderers: Required<ChangelogRenderers> = {
  renderHeader: _renderHeader,
  renderChangeTypeSection: _renderChangeTypeSection,
  renderChangeTypeHeader: _renderChangeTypeHeader,
  renderEntries: _renderEntries,
  renderEntry: _renderEntry,
};

export function renderPackageChangelog(renderInfo: PackageChangelogRenderInfo) {
  const { renderHeader, renderChangeTypeSection } = renderInfo.renderers;
  const versionHeader = renderHeader(renderInfo);

  return [
    versionHeader,
    renderChangeTypeSection('major', renderInfo),
    renderChangeTypeSection('minor', renderInfo),
    renderChangeTypeSection('patch', renderInfo),
    renderChangeTypeSection('prerelease', renderInfo),
  ]
    .filter(Boolean)
    .join('\n\n');
}

function _renderHeader(renderInfo: PackageChangelogRenderInfo): string {
  return `## ${renderInfo.newVersionChangelog.version}\n\n${renderInfo.newVersionChangelog.date.toUTCString()}`;
}

function _renderChangeTypeSection(changeType: ChangeType, renderInfo: PackageChangelogRenderInfo): string {
  const { renderChangeTypeHeader, renderEntries } = renderInfo.renderers;
  const entries = renderInfo.newVersionChangelog.comments[changeType];
  return entries && entries.length
    ? `${renderChangeTypeHeader(changeType, renderInfo)}\n\n${renderEntries(changeType, renderInfo)}`
    : '';
}

function _renderChangeTypeHeader(changeType: ChangeType, renderInfo: PackageChangelogRenderInfo): string {
  return `### ${groupNames[changeType]}`;
}

function _renderEntries(changeType: ChangeType, renderInfo: PackageChangelogRenderInfo): string {
  const entries = renderInfo.newVersionChangelog.comments[changeType];
  if (!entries || !entries.length) {
    return '';
  }
  const { renderEntry } = renderInfo.renderers;
  if (renderInfo.isGrouped) {
    const entriesMap = _.groupBy(entries, entry => entry.package);

    return Object.keys(entriesMap)
      .map(pkgName => {
        const entriesText = entriesMap[pkgName].map(entry => `  ${renderEntry(entry, renderInfo)}`).join('\n');
        return `- \`${pkgName}\`\n${entriesText}`;
      })
      .join('\n');
  }

  return entries.map(entry => renderEntry(entry, renderInfo)).join('\n');
}

function _renderEntry(entry: ChangelogEntry, renderInfo: PackageChangelogRenderInfo): string {
  return `- ${entry.comment} (${entry.author})`;
}
