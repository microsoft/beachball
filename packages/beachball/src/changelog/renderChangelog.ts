import { PackageChangelog } from '../types/ChangeLog';
import { renderPackageChangelog, defaultRenderers } from './renderPackageChangelog';
import { ChangelogOptions, PackageChangelogRenderInfo } from '../types/ChangelogOptions';

export interface MarkdownChangelogRenderOptions extends Omit<PackageChangelogRenderInfo, 'renderers'> {
  previousContent: string;
  changelogOptions: ChangelogOptions;
}

export const markerComment = '<!-- Start content -->';

export async function renderChangelog(renderOptions: MarkdownChangelogRenderOptions): Promise<string> {
  const {
    previousJson,
    previousContent = '',
    newVersionChangelog,
    isGrouped,
    changelogOptions: { renderPackageChangelog: customRenderPackageChangelog, customRenderers },
  } = renderOptions;

  let previousLogEntries: string;
  if (previousContent.includes(markerComment)) {
    // Preferably determine where the previous entries start based on a special comment
    previousLogEntries = previousContent.split(markerComment, 2)[1].trim();
  } else {
    // Otherwise look for an h2 (used as version header with default renderer).
    // If that's not present, preserve the previous content as-is.
    const h2Match = previousContent.match(/^## /m);
    previousLogEntries = h2Match ? previousContent.substring(h2Match.index!) : previousContent;
  }

  try {
    if (customRenderPackageChangelog || customRenderers) {
      console.log('Using custom renderer for package version changelog.');
    }

    const renderInfo: PackageChangelogRenderInfo = {
      previousJson,
      newVersionChangelog,
      isGrouped,
      renderers: {
        ...defaultRenderers,
        ...customRenderers,
      },
    };

    return (
      [
        renderChangelogHeader(newVersionChangelog),
        markerComment,
        await (customRenderPackageChangelog || renderPackageChangelog)(renderInfo),
        previousLogEntries,
      ]
        .join('\n\n')
        .trim() + '\n'
    );
  } catch (err) {
    console.log('Error occurred rendering package version changelog:', err);
    return '';
  }
}

function renderChangelogHeader(changelog: PackageChangelog): string {
  return (
    `# Change Log - ${changelog.name}\n\n` +
    `This log was last generated on ${changelog.date.toUTCString()} and should not be manually modified.`
  );
}
