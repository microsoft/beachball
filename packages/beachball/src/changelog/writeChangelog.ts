import path from 'path';
import fs from 'fs-extra';
import _ from 'lodash';
import { ChangeSet } from '../types/ChangeInfo';
import { PackageInfo } from '../types/PackageInfo';
import { getPackageChangelogs } from './getPackageChangelogs';
import { renderChangelog } from './renderChangelog';
import { renderJsonChangelog } from './renderJsonChangelog';
import { BeachballOptions } from '../types/BeachballOptions';
import { isPathIncluded } from '../monorepo/utils';
import { PackageChangelog, ChangelogJson } from '../types/ChangeLog';
import { mergeChangelogs } from './mergeChangelogs';

export async function writeChangelog(
  options: BeachballOptions,
  changeSet: ChangeSet,
  packageInfos: {
    [pkg: string]: PackageInfo;
  }
): Promise<void> {
  const groupedChangelogPaths = await writeGroupedChangelog(options, changeSet, packageInfos);
  const groupedChangelogPathSet = new Set(groupedChangelogPaths);

  const changelogs = getPackageChangelogs(changeSet, packageInfos);
  // Use a standard for loop here to prevent potentially firing off multiple network requests at once
  // (in case any custom renderers have network requests)
  for (const pkg of Object.keys(changelogs)) {
    const packagePath = path.dirname(packageInfos[pkg].packageJsonPath);
    if (groupedChangelogPathSet?.has(packagePath)) {
      console.log(`Changelog for ${pkg} has been written as a group here: ${packagePath}`);
    } else {
      await writeChangelogFiles(options, changelogs[pkg], packagePath, false);
    }
  }
}

async function writeGroupedChangelog(
  options: BeachballOptions,
  changeSet: ChangeSet,
  packageInfos: {
    [pkg: string]: PackageInfo;
  }
): Promise<string[]> {
  if (!options.changelog) {
    return [];
  }

  const { groups: changelogGroups } = options.changelog;
  if (!changelogGroups || changelogGroups.length < 1) {
    return [];
  }

  const changelogs = getPackageChangelogs(changeSet, packageInfos);
  const groupedChangelogs: {
    [path: string]: { changelogs: PackageChangelog[]; masterPackage: PackageInfo };
  } = {};

  for (const pkg in changelogs) {
    const packagePath = path.dirname(packageInfos[pkg].packageJsonPath);
    const relativePath = path.relative(options.path, packagePath);
    for (const group of changelogGroups) {
      const { changelogPath, masterPackageName } = group;
      const masterPackage = packageInfos[masterPackageName];
      if (!masterPackage) {
        console.warn(`master package ${masterPackageName} does not exist.`);
        continue;
      }
      if (!fs.existsSync(changelogPath)) {
        console.warn(`changelog path ${changelogPath} does not exist.`);
        continue;
      }

      const isInGroup = isPathIncluded(relativePath, group.include, group.exclude);
      if (isInGroup) {
        if (!groupedChangelogs[changelogPath]) {
          groupedChangelogs[changelogPath] = {
            changelogs: [],
            masterPackage,
          };
        }

        groupedChangelogs[changelogPath].changelogs.push(changelogs[pkg]);
      }
    }
  }

  const changelogAbsolutePaths: string[] = [];
  for (const changelogPath in groupedChangelogs) {
    const { masterPackage, changelogs } = groupedChangelogs[changelogPath];
    const groupedChangelog = mergeChangelogs(changelogs, masterPackage);
    if (groupedChangelog) {
      await writeChangelogFiles(options, groupedChangelog, changelogPath, true);
      changelogAbsolutePaths.push(path.resolve(changelogPath));
    }
  }

  return changelogAbsolutePaths;
}

async function writeChangelogFiles(
  options: BeachballOptions,
  newVersionChangelog: PackageChangelog,
  changelogPath: string,
  isGrouped: boolean
): Promise<void> {
  let previousJson: ChangelogJson | undefined;

  // Update CHANGELOG.json
  const changelogJsonFile = path.join(changelogPath, 'CHANGELOG.json');
  try {
    previousJson = fs.existsSync(changelogJsonFile) ? fs.readJSONSync(changelogJsonFile) : undefined;
  } catch (e) {
    console.warn('CHANGELOG.json is invalid:', e);
  }
  try {
    const nextJson = renderJsonChangelog(newVersionChangelog, previousJson);
    fs.writeJSONSync(changelogJsonFile, nextJson, { spaces: 2 });
  } catch (e) {
    console.warn('Problem writing to CHANGELOG.json:', e);
  }

  // Update CHANGELOG.md
  if (
    newVersionChangelog.comments.major ||
    newVersionChangelog.comments.minor ||
    newVersionChangelog.comments.patch ||
    newVersionChangelog.comments.prerelease
  ) {
    const changelogFile = path.join(changelogPath, 'CHANGELOG.md');
    const previousContent = fs.existsSync(changelogFile) ? fs.readFileSync(changelogFile).toString() : '';

    const newChangelog = await renderChangelog({
      previousJson,
      previousContent,
      newVersionChangelog,
      isGrouped,
      changelogOptions: options.changelog || {},
    });

    fs.writeFileSync(changelogFile, newChangelog);
  }
}
