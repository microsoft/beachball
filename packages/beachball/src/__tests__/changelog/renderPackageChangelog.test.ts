import { PackageChangelogRenderInfo } from '../../types/ChangelogOptions';
import { defaultRenderers, renderPackageChangelog } from '../../changelog/renderPackageChangelog';

const { renderEntry, renderEntries, renderChangeTypeHeader, renderChangeTypeSection, renderHeader } = defaultRenderers;

const leadingNewlineRegex = /^\n/;
const trailingNewlineRegex = /\n$/;

describe('changelog renderers -', () => {
  function getRenderInfo(): PackageChangelogRenderInfo {
    return {
      isGrouped: false,
      newVersionChangelog: {
        date: new Date('Thu Aug 22 2019 14:20:40 GMT-0700 (Pacific Daylight Time)'),
        name: 'foo',
        tag: 'foo_v1.2.3',
        version: '1.2.3',
        comments: {
          major: [],
          minor: [
            { comment: 'Awesome change', author: 'user1@example.com', commit: 'sha1', package: 'foo' },
            { comment: 'Boring change', author: 'user2@example.com', commit: 'sha2', package: 'foo' },
          ],
          patch: [
            { comment: 'Fix', author: 'user1@example.com', commit: 'sha3', package: 'foo' },
            { comment: 'stuff', author: 'user2@example.com', commit: 'sha4', package: 'foo' },
          ],
        },
      },
      previousJson: {} as any,
      renderers: { ...defaultRenderers }, // copy in case of modification
    };
  }

  function getGroupedRenderInfo(): PackageChangelogRenderInfo {
    const renderInfo = getRenderInfo();
    renderInfo.isGrouped = true;
    renderInfo.newVersionChangelog.comments.minor![0].package = 'bar';
    renderInfo.newVersionChangelog.comments.patch![1].package = 'bar';
    return renderInfo;
  }

  function doBasicTests(result: string) {
    expect(result).not.toMatch(leadingNewlineRegex);
    expect(result).not.toMatch(trailingNewlineRegex);
    expect(result).toMatchSnapshot();
  }

  describe('renderEntry', () => {
    it('has correct output', () => {
      const renderInfo = getRenderInfo();
      const result = renderEntry(renderInfo.newVersionChangelog.comments.minor![0], renderInfo);
      doBasicTests(result);
    });

    it('has correct grouped output', () => {
      const renderInfo = getGroupedRenderInfo();
      const result = renderEntry(renderInfo.newVersionChangelog.comments.minor![0], renderInfo);
      doBasicTests(result);
    });
  });

  describe('renderEntries', () => {
    it('has correct output', () => {
      const renderInfo = getRenderInfo();
      const result = renderEntries('minor', renderInfo);
      doBasicTests(result);
    });

    it('has correct grouped output', () => {
      const renderInfo = getGroupedRenderInfo();
      const result = renderEntries('minor', renderInfo);
      doBasicTests(result);
    });
  });

  describe('renderChangeTypeHeader', () => {
    it('has correct output', () => {
      const renderInfo = getRenderInfo();
      const result = renderChangeTypeHeader('minor', renderInfo);
      doBasicTests(result);
    });

    it('has correct grouped output', () => {
      const renderInfo = getGroupedRenderInfo();
      const result = renderChangeTypeHeader('minor', renderInfo);
      doBasicTests(result);
    });
  });

  describe('renderChangeTypeSection', () => {
    it('has correct output', () => {
      const renderInfo = getRenderInfo();
      const result = renderChangeTypeSection('minor', renderInfo);
      doBasicTests(result);
    });

    it('has correct grouped output', () => {
      const renderInfo = getGroupedRenderInfo();
      const result = renderChangeTypeSection('minor', renderInfo);
      doBasicTests(result);
    });
  });

  describe('renderHeader', () => {
    it('has correct output', () => {
      const renderInfo = getRenderInfo();
      const result = renderHeader(renderInfo);
      doBasicTests(result);
    });

    it('has correct grouped output', () => {
      const renderInfo = getGroupedRenderInfo();
      const result = renderHeader(renderInfo);
      doBasicTests(result);
    });
  });

  describe('renderPackageChangelog', () => {
    it('has correct output', () => {
      const renderInfo = getRenderInfo();
      const result = renderPackageChangelog(renderInfo);
      doBasicTests(result);
    });

    it('has correct grouped output', () => {
      const renderInfo = getGroupedRenderInfo();
      const result = renderPackageChangelog(renderInfo);
      doBasicTests(result);
    });

    it('uses custom renderEntry', () => {
      const renderInfo = getGroupedRenderInfo();
      renderInfo.renderers.renderEntry = (entry, renderInfo) => `- ${entry.comment} (#123)`;

      const result = renderPackageChangelog(renderInfo);
      expect(result).toContain('#123');
      expect(result).toMatchSnapshot();
    });

    it('uses custom renderEntries', () => {
      const renderInfo = getRenderInfo();
      renderInfo.renderers.renderEntries = (changeType, renderInfo) => {
        const entries = renderInfo.newVersionChangelog.comments[changeType];
        return entries ? entries.map(entry => `${entry.comment}!!!`).join('\n\n') : '';
      };

      const result = renderPackageChangelog(renderInfo);
      expect(result).toContain('!!!');
      expect(result).toMatchSnapshot();
    });

    it('uses custom renderChangeTypeHeader', () => {
      const renderInfo = getRenderInfo();
      renderInfo.renderers.renderChangeTypeHeader = (changeType, renderInfo) =>
        changeType === 'minor' || changeType === 'major' ? '### Important stuff' : '### Boring stuff';

      const result = renderPackageChangelog(renderInfo);
      expect(result).toContain('### Important stuff');
      expect(result).toMatchSnapshot();
    });

    it('uses custom renderChangeTypeSection', () => {
      const renderInfo = getRenderInfo();
      renderInfo.renderers.renderChangeTypeSection = (changeType, renderInfo) =>
        changeType === 'minor' || changeType === 'major' ? renderChangeTypeSection(changeType, renderInfo) : '';

      const result = renderPackageChangelog(renderInfo);
      expect(result).not.toContain('Patches');
      expect(result).toMatchSnapshot();
    });

    it('uses custom renderHeader', () => {
      const renderInfo = getRenderInfo();
      renderInfo.renderers.renderHeader = renderInfo =>
        [
          `## ${renderInfo.newVersionChangelog.version}`,
          renderInfo.newVersionChangelog.date.toUTCString(),
          `[Compare changes](http://real-github-compare-link)`,
        ].join('\n');

      const result = renderPackageChangelog(renderInfo);
      expect(result).toContain('Compare changes');
      expect(result).toMatchSnapshot();
    });
  });
});
