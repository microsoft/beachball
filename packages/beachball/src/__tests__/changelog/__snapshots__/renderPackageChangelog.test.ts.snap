// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changelog renderers - renderChangeTypeHeader has correct grouped output 1`] = `"### Minor changes"`;

exports[`changelog renderers - renderChangeTypeHeader has correct output 1`] = `"### Minor changes"`;

exports[`changelog renderers - renderChangeTypeSection has correct grouped output 1`] = `
"### Minor changes

- \`bar\`
  - Awesome change (user1@example.com)
- \`foo\`
  - Boring change (user2@example.com)"
`;

exports[`changelog renderers - renderChangeTypeSection has correct output 1`] = `
"### Minor changes

- Awesome change (user1@example.com)
- Boring change (user2@example.com)"
`;

exports[`changelog renderers - renderEntries has correct grouped output 1`] = `
"- \`bar\`
  - Awesome change (user1@example.com)
- \`foo\`
  - Boring change (user2@example.com)"
`;

exports[`changelog renderers - renderEntries has correct output 1`] = `
"- Awesome change (user1@example.com)
- Boring change (user2@example.com)"
`;

exports[`changelog renderers - renderEntry has correct grouped output 1`] = `"- Awesome change (user1@example.com)"`;

exports[`changelog renderers - renderEntry has correct output 1`] = `"- Awesome change (user1@example.com)"`;

exports[`changelog renderers - renderHeader has correct grouped output 1`] = `
"## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT"
`;

exports[`changelog renderers - renderHeader has correct output 1`] = `
"## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT"
`;

exports[`changelog renderers - renderPackageChangelog has correct grouped output 1`] = `
"## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT

### Minor changes

- \`bar\`
  - Awesome change (user1@example.com)
- \`foo\`
  - Boring change (user2@example.com)

### Patches

- \`foo\`
  - Fix (user1@example.com)
- \`bar\`
  - stuff (user2@example.com)"
`;

exports[`changelog renderers - renderPackageChangelog has correct output 1`] = `
"## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT

### Minor changes

- Awesome change (user1@example.com)
- Boring change (user2@example.com)

### Patches

- Fix (user1@example.com)
- stuff (user2@example.com)"
`;

exports[`changelog renderers - renderPackageChangelog uses custom renderChangeTypeHeader 1`] = `
"## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT

### Important stuff

- Awesome change (user1@example.com)
- Boring change (user2@example.com)

### Boring stuff

- Fix (user1@example.com)
- stuff (user2@example.com)"
`;

exports[`changelog renderers - renderPackageChangelog uses custom renderChangeTypeSection 1`] = `
"## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT

### Minor changes

- Awesome change (user1@example.com)
- Boring change (user2@example.com)"
`;

exports[`changelog renderers - renderPackageChangelog uses custom renderEntries 1`] = `
"## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT

### Minor changes

Awesome change!!!

Boring change!!!

### Patches

Fix!!!

stuff!!!"
`;

exports[`changelog renderers - renderPackageChangelog uses custom renderEntry 1`] = `
"## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT

### Minor changes

- \`bar\`
  - Awesome change (#123)
- \`foo\`
  - Boring change (#123)

### Patches

- \`foo\`
  - Fix (#123)
- \`bar\`
  - stuff (#123)"
`;

exports[`changelog renderers - renderPackageChangelog uses custom renderHeader 1`] = `
"## 1.2.3
Thu, 22 Aug 2019 21:20:40 GMT
[Compare changes](http://real-github-compare-link)

### Minor changes

- Awesome change (user1@example.com)
- Boring change (user2@example.com)

### Patches

- Fix (user1@example.com)
- stuff (user2@example.com)"
`;
